test/unified-config.test.js

```diff
@@ -0,0 +1,453 @@
+    const config = {
+      transactionSettings: [
+        {type: 'url', regex: '/xy(zzy', tracing: 'disabled'}
+      ]
+    };
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const msg = 'Invalid regular expression: //xy(zzy/: Unterminated group';
+    const settingsErrors = [toTransactionSettingsError(config.transactionSettings[0], msg)];
+    doChecks(cfg, {settingsErrors});
+  })
+
+  it('should allow a RegExp in module-based x-action settings', function () {
+    const config = {transactionSettings: [
+      {type: 'url', regex: /xyzzy/, tracing: 'disabled'},
+      {type: 'url', regex: 'hello', tracing: 'disabled'},
+      {type: 'url', string: '/xy(zzy', tracing: 'disabled'},
+    ]};
+    const literal = [
+      'module.exports = {transactionSettings: [',
+      '  {type: "url", regex: /xyzzy/, tracing: "disabled"},',
+      '  {type: "url", regex: "hello", tracing: "disabled"},',
+      '  {type: "url", string: "/xy(zzy", tracing: "disabled"},',
+      ']}', ''
+    ];
+    writeConfigJs(literal.join('\n'));
+
+    // specify the filename with extension to work around node bug/feature/issue.
+    const file = process.env.APPOPTICS_APM_CONFIG_NODE = 'appoptics-apm.js';
+
+    const cfg = guc();
+
+    const transactionSettings = toInternalTransactionSettings(config.transactionSettings);
+    doChecks(cfg, {file: `${process.cwd()}/${file}`, transactionSettings});
+  })
+
+  it('should report invalid transactionSettings entries', function () {
+    const config = {transactionSettings: [
+      {type: 'z'},
+      {regex: 17},
+      {type: 'url', regex: 'hello'},
+      {type: 'url', string: 'hello', tracing: 'invalid'},
+      {type: 'url', tracing: 'enabled'},
+      {type: 'url', regex: 'not-real-regex', string: 'i am a string'},
+    ]}
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const settingsErrors = [
+      toTransactionSettingsError(config.transactionSettings[0], 'invalid type: "z"'),
+      toTransactionSettingsError(config.transactionSettings[1], 'invalid type: "undefined"'),
+      toTransactionSettingsError(config.transactionSettings[2], 'invalid tracing value: "undefined"'),
+      toTransactionSettingsError(config.transactionSettings[3], 'invalid tracing value: "invalid"'),
+      toTransactionSettingsError(config.transactionSettings[4], 'must specify one, not both, of "string" and "regex"'),
+      toTransactionSettingsError(config.transactionSettings[5], 'must specify one, not both, of "string" and "regex"'),
+    ];
+
+    doChecks(cfg, {settingsErrors});
+  })
+
+  it('should allow a single transactionSettings entry', function () {
+    const config = {transactionSettings: {type: 'url', string: 'i\'m a shrimp', tracing: 'disabled'}};
+    writeConfigJSON(config);
+
+    const cfg = guc();
+
+    const transactionSettings = toInternalTransactionSettings(config.transactionSettings);
+    doChecks(cfg, {transactionSettings});
+  })
+
+})
```

Comment:
very thorough!
The only additional test I could think of is when tracing is disabled and there is a setting to enable certain urls. This would enable dowample and dometrics.

Reply:
i don't think that should be covered by this test. the designed goal of this module is to process the configuration file and env vars in a consistent and predictable priority. getting into the semantics of what the properties mean is beyond the scope. for example, the module doesn't verify that a the serviceKey is formatted correctly - that is handled by the caller (index.js).

make sense? or am i off base with my thinking?

Reply:
My thinking about tests is that they should discover if some refactoring goes wrong. So I would like to have a test that goes through the enabled case, because there is a tiny bit of code that is different from the disabled case.
```
if (setting.tracing === 'enabled') {
      goodEntry.doSample = goodEntry.doMetrics = true
    }
```

Reply:
i added the coverage on line 370 (the "should allow JSON x-action settings" test). no real reason for a new test (i don't think). thanks!